File: python/removing_stars_from_a_string.py
class Solution:
    def removeStars(self, s: str) -> str:
        """time complexity: O(n), space complexity: O(n)"""
        stack: list[str] = []
        for c in s:
            if len(stack) >= 0 and c == "*":
                stack.pop()
            else:
                stack.append(c)
        return "".join(stack)


def test_removeStars_case_1():
    # arrange
    s: str = "leet**cod*e"
    expected: str = "lecoe"

    # act
    solution = Solution()
    actual = solution.removeStars(s)

    # assert
    assert expected == actual


def test_removeStars_case_2():
    # arrange
    s: str = "erase*****"
    expected: str = ""

    # act
    solution = Solution()
    actual = solution.removeStars(s)

    # assert
    assert expected == actual

File: rust/src/code/removing_stars_from_a_string.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity: O(n)
    pub fn remove_stars(s: String) -> String {
        let mut stack: Vec<char> = Vec::new();
        let s: Vec<char> = s.chars().collect();
        for i in 0..s.len() {
            if s[i] == '*' {
                stack.pop().unwrap();
            } else {
                stack.push(s[i]);
            }
        }

        stack.iter().collect::<String>()
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_remove_stars_case_1() {
        // arrange
        let s: String = "leet**cod*e".to_string();
        let expected: String = "lecoe".to_string();

        // act
        let actual = Solution::remove_stars(s);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_remove_stars_case_2() {
        // arrange
        let s: String = "erase*****".to_string();
        let expected: String = "".to_string();

        // act
        let actual = Solution::remove_stars(s);

        // assert
        assert_eq!(expected, actual);
    }
}
